File: src/itav_agv_controller.cpp
#include "itav_agv_controller/itav_agv_controller.hpp"


ITAVController::ITAVController() : Node("itav_agv_controller")
{
    RCLCPP_INFO(this->get_logger(), "controller initialized");

    odom_sub = create_subscription<nav_msgs::msg::Odometry>(
        "itav_agv/odometry/filtered",
        10,
        std::bind(&ITAVController::odom_callback, this, std::placeholders::_1));
    tracked_pose = create_subscription<geometry_msgs::msg::PoseStamped>(
        "itav_agv/tracked_pose",
        10,
        std::bind(&ITAVController::pose_callback, this, std::placeholders::_1));
    goal_sub = create_subscription<geometry_msgs::msg::PoseStamped>(
        "goal_pose",
        10,
        std::bind(&ITAVController::goal_callback, this, std::placeholders::_1));
    cmd_vel_pub = create_publisher<geometry_msgs::msg::Twist>("itav_agv/cmd_vel", 10);
    trajectory_pub = create_publisher<nav_msgs::msg::Path>("trajectory", 10);
    grid_map_sub = create_subscription<nav_msgs::msg::OccupancyGrid>(
        "projected_map",
        10,
        std::bind(&ITAVController::grid_map_callback, this, std::placeholders::_1));
    obstacle_sub = create_subscription<std_msgs::msg::Float32MultiArray>(
        "obstacles_state",
        10,
        std::bind(&ITAVController::obstacle_callback, this, std::placeholders::_1));
    ob_pub = create_publisher<visualization_msgs::msg::MarkerArray>("obstacles", 10);

    // dwa = DWAPlanner(current_pose);
    // trajectory.poses.push_back(geometry_msgs::msg::PoseStamped());
}

void ITAVController::obstacle_callback(const std_msgs::msg::Float32MultiArray::SharedPtr msg)
{
    auto obstacles = msg->data;
    std::vector<double> obs(obstacles.size());
    std::copy(obstacles.begin(), obstacles.end(), obs.begin());
    auto obstacles_matrix = Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>(obs.data(), obstacles.size() / 4, 4);
    if (current_pose.size() == 3) {
        double robot_x = current_pose[0];
        double robot_y = current_pose[1];
        obstacles_matrix = obstacles_matrix.rowwise().squaredNorm().eval().array() > 0.75 * 0.75;
        obstacles_matrix = (obstacles_matrix.col(0).array() + robot_x).matrix().transpose().finished().eval().comma(obstacles_matrix.col(1).array() + robot_y);
    }
    dwa.config.ob = obstacles_matrix;
}

void ITAVController::grid_map_callback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
{
    grid.clear();
    auto map_data = msg->data;
    int map_width = msg->info.width;
    int map_height = msg->info.height;
    double resolution = msg->info.resolution;
    double origin_x = msg->info.origin.position.x;
    double origin_y = msg->info.origin.position.y;

    std::vector<int> map_row(map_width, 0);
    for (int i = 0; i < map_height; i++) {
        auto row_start = map_data.begin() + i * map_width;
        auto row_end = row_start + map_width
    std::vector<int> map_row(row_start, row_end);
    grid.push_back(map_row);
}

auto map_to_world = [&](const int& i, const int& j) {
    return std::make_pair(origin_x + j * resolution, origin_y + i * resolution);
};

std::vector<Eigen::VectorXd> obstacles;
for (int i = 0; i < grid.size(); i++) {
    for (int j = 0; j < grid[i].size(); j++) {
        if (grid[i][j] > 50) {
            auto world_pos = map_to_world(i, j);
            obstacles.push_back(Eigen::VectorXd::Map(world_pos.data(), world_pos.size()));
        }
    }
}

if (obstacles.size() > 0) {
    Eigen::MatrixXd obs(obstacles.size(), 4);
    for (int i = 0; i < obstacles.size(); i++) {
        obs.row(i) << obstacles[i](0), obstacles[i](1), 0, 0;
    }
    dwa.config.ob = obs;
}
}

void ITAVController::goal_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
{
auto pose = msg->pose;
double x = pose.position.x;
double y = pose.position.y;
double _, , yaw;
tf2::Quaternion q;
tf2::fromMsg(pose.orientation, q);
tf2::Matrix3x3(q).getRPY(, _, yaw);
RCLCPP_INFO(this->get_logger(), "goal_pose received (%f, %f, %f)", x, y, yaw);
goal_pose = { x, y, yaw };
dwa.set_goal_pose(goal_pose);
}

void ITAVController::odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
{
auto pose = msg->pose.pose;
double x = pose.position.x;
double y = pose.position.y;
double _, , yaw;
tf2::Quaternion q;
tf2::fromMsg(pose.orientation, q);
tf2::Matrix3x3(q).getRPY(, _, yaw);
current_pose = { x, y, yaw };
dwa.set_current_pose(current_pose);
}

void ITAVController::pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
{
auto pose = msg->pose;
double x = pose.position.x;
double y = pose.position.y;
double _, , yaw;
tf2::Quaternion q;
tf2::fromMsg(pose.orientation, q);
tf2::Matrix3x3(q).getRPY(, _, yaw);
current_pose = { x, y, yaw };
dwa.set_current_pose(current_pose);
}

void ITAVController::controller()
{
if (goal_pose.size() == 3 && current_pose.size() == 3) {
auto u_and_predicted_trajectory = dwa.dwa_control();
auto u = u_and_predicted_trajectory.first;
auto predicted_trajectory = u_and_predicted_trajectory.second;
dwa.set_cmd_vel(u);

    trajectory.poses.clear();
    for (auto& pose : predicted_trajectory) {
        geometry_msgs::msg::PoseStamped p;
        p.header.frame_id = "map";
        p.pose.position.x = pose(0);
        p.pose.position.y = pose(1);
        p.pose.position.z = 0;
        tf2::Quaternion q;
        q.setRPY(0, 0, pose(2));
        p.pose.orientation = tf2::toMsg(q);
        trajectory.poses.push_back(p);
    }
    trajectory.header.frame_id = "map";
    trajectory_pub->publish(trajectory);
    auto vel = geometry_msgs::msg::Twist();
    vel.linear.x = u(0);
    vel.angular.z = u(1);
    RCLCPP_INFO(this->get_logger(), "vel: (%f, %f)", vel.linear.x, vel.angular.z);
    RCLCPP_INFO(this->get_logger(), "goal_pose: (%f, %f, %f)", goal_pose[0], goal_pose[1], goal_pose[2]);
    publish_vel(vel);

    double dist_to_goal = std::hypot(goal_pose[0] - current_pose[0], goal_pose[1] - current_pose[1]);
    if (dist_to_goal < 0.5) {
        RCLCPP_INFO(this->get_logger(), "goal reached");
        goal_pose.clear();
        publish_vel(geometry_msgs::msg::Twist());
    }
}
else {
    RCLCPP_INFO(this->get_logger(), "waiting for goal_pose");
}
}

void ITAVController::plotter()
{
if (goal_pose.size() == 3 && current_pose.size() == 3) {
dwa.plot_all();
}
}

void ITAVController::publish_vel(const geometry_msgs::msg::Twist& vel)
{
cmd_vel_pub->publish(vel);
}

int main(int argc, char** argv)
{
rclcpp::init(argc, argv);
auto node = std::make_shared<ITAVController>(rclcpp::NodeOptions());
RCLCPP_INFO(node->get_logger(), "waiting for goal_pose");
rclcpp::spin(node);
rclcpp::shutdown();
return 0;
}

File: scripts/itav_agv_controller.py
#! /usr/bin/env python3

import utils.dwa
import rclpy
from tf_transformations import euler_from_quaternion, quaternion_from_euler
import math
import numpy as np
import matplotlib.pyplot as plt

from geometry_msgs.msg import Twist, PoseStamped, Point, PoseWithCovarianceStamped
from std_msgs.msg import Float32MultiArray
from nav_msgs.msg import Odometry, Path, OccupancyGrid
from visualization_msgs.msg import MarkerArray, Marker
from nav_msgs.srv import GetMap
from cartographer_ros_msgs.srv import StartTrajectory
class ITAVController:
  def __init__(self, node):
    print("controller initialized")
    self.current_pose = None
    self.goal_pose = None
    self.grid = None
    self.odom_sub = node.create_subscription(Odometry, 'itav_agv/odometry/filtered', self.odom_callback, 10)
    self.tracked_pose = node.create_subscription(PoseWithCovarianceStamped, 'itav_agv/global_pose', self.pose_callback, 10)
    self.goal_sub = node.create_subscription(PoseStamped, 'goal_pose', self.goal_callback, 10)
    self.cmd_vel_pub = node.create_publisher(Twist, 'itav_agv/cmd_vel', 10)
    self.trajectory_pub = node.create_publisher(Path, 'trajectory1', 10)
    # self.grid_map_sub = node.create_subscription(OccupancyGrid, '/global_costmap/costmap', self.grid_map_callback, 10)
    self.grid_map_sub = node.create_subscription(OccupancyGrid, 'map', self.grid_map_callback, 1)
    self.obstacle_sub = node.create_subscription(Float32MultiArray, 'obstacles_state', self.obstacle_callback, 10)
    self.ob_pub = node.create_publisher(MarkerArray, 'obstacles', 10)
    # get map service call
    self.get_map_client = node.create_client(GetMap, '/map_server/map')
    # cartographer start trajectory service call
    self.start_trajectory_client = node.create_client(StartTrajectory, '/start_trajectory')
    # list of trajectories publisher
    self.trajectories_pub = node.create_publisher(MarkerArray, 'trajectory_list1', 10)
    
    self.dwa = utils.dwa.DWA(self.current_pose)
    self.trajectory = np.array([[0,0,0]])

    # obstacles
    self.dwa.config.ob = np.array([[20, 20, 0, 0.2,]
                            ]) 



    while not self.get_map_client.wait_for_service(timeout_sec=1.0):
      print('map ervice not available, waiting again...')
    self.req = GetMap.Request()
    self.future = self.get_map_client.call_async(self.req)
    rclpy.spin_until_future_complete(node, self.future)
    if self.future.result() is not None:
      self.grid = np.array(self.future.result().map.data).reshape(self.future.result().map.info.width, self.future.result().map.info.height).T
      self.dwa.set_map(self.grid)
      print("map received")
    
    # while not self.start_trajectory_client.wait_for_service(timeout_sec=1.0) or self.current_pose is None:
    #   print('cartographer service not available, waiting again...', self.current_pose)
    #   self.current_pose = [20.0, 55.0, 0.0]
    # self.req = StartTrajectory.Request()
    # self.req.configuration_directory = "/home/phoenix/ros2_ws/src/tavil/itav_agv_bringup/config"
    # self.req.configuration_basename = "cartographer_2d.lua"
    # self.req.use_initial_pose = bool(0)
    # self.req.relative_to_trajectory_id = 0
    # self.req.initial_pose.position.x = self.current_pose[0]
    # self.req.initial_pose.position.y = self.current_pose[1]
    # self.req.initial_pose.position.z = 0.0
    # self.req.initial_pose.orientation.x = 0.0
    # self.req.initial_pose.orientation.y = 0.0
    # self.req.initial_pose.orientation.z = 0.0
    # self.req.initial_pose.orientation.w = 1.0
    # self.future = self.start_trajectory_client.call_async(self.req)
    # rclpy.spin_until_future_complete(node, self.future)
    # if self.future.result() is not None:
    #   print("trajectory started")

    # ros2 timer controller
    self.timer = node.create_timer(0.1, self.controller)
    # self.timer = node.create_timer(0.5, self.plotter)

  def obstacle_callback(self, msg):
    # pass
    # first remove the obstacle that is the same as the current robot pose by checking the distance
    print("obstacles size", len(msg.data))
    if len(msg.data) == 0:
       obstacles = np.array([[800, 800, 0, 0.2,], [800, 800, 0, 0.2,], [800, 800, 0, 0.2,]])
    else:
      obstacles = np.array(msg.data).reshape(-1, 4)
      # remove obstacles that are far from current pose
      if self.current_pose is not None:
        obstacles = obstacles[np.linalg.norm(obstacles[:, :2] - self.current_pose[:2], axis=1) < 70]
        obstacles = np.append(obstacles, np.array([[800, 800, 0, 0.2,], [800, 800, 0, 0.2,], [800, 800, 0, 0.2,]]), axis=0)
      # if self.grid is not None:
      #   obstacles = obstacles[self.grid[20*obstacles[:, 0].astype(int), 20*obstacles[:, 1].astype(int)] == 0]
    if self.current_pose is not None:
      d = np.hypot(self.current_pose[0], self.current_pose[1])
      r = 1. 
      x = np.zeros(2)
      x[0] = self.current_pose[0] + r * np.cos(self.current_pose[2])
      x[1] = self.current_pose[1] + r * np.sin(self.current_pose[2])
      obstacles = obstacles[np.linalg.norm(obstacles[:, :2] - x, axis=1) > 0.75]
    self.dwa.config.ob = obstacles
    self.publish_ob()

  def grid_map_callback(self, msg):
    # print("grid map shape", msg.info.width, msg.info.height)
    self.grid = np.array(msg.data).reshape(msg.info.width, msg.info.height)
    origin = np.array([msg.info.origin.position.x, msg.info.origin.position.y])
    # origin adjustment
    self.grid = np.roll(self.grid, int(origin[0]), axis=0)
    self.grid = np.roll(self.grid, int(origin[1]), axis=1)
    print("grid map shape", self.grid.shape)
    self.dwa.set_map(self.grid)
  # rviz goal_pose subscriber
  def goal_callback(self, msg):

    x = msg.pose.position.x
    y = msg.pose.position.y
    _,_,yaw = euler_from_quaternion([msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w])
    print("goal_pose received", x, y, yaw)
    self.goal_pose = [x, y, yaw]
    self.dwa.set_goal_pose(self.goal_pose)
    # publish ob as marker array
    
  def publish_ob(self):
    # publish obstacles as marker array
    marker_array = MarkerArray()
    vel_marker_array = MarkerArray()
    for i in range(len(self.dwa.config.ob)):
        marker = Marker()
        vel_marker = Marker()
        marker.header.frame_id = "odom"
        # marker.header.stamp = rclpy.get_clock().now().to_msg()
        # u = [np.random.uniform(-0.75, 0.75), np.random.uniform(-0.5, 0.5)]
        # x = self.dwa.motion(self.dwa.config.ob[i,:], u, 1)
        # self.dwa.config.ob[i,:] = x
        marker.ns = "obstacles"
        marker.id = i
        marker.type = Marker.CUBE
        marker.action = Marker.ADD
        marker.pose.position.x = float(self.dwa.config.ob[i,0])
        marker.pose.position.y = float(self.dwa.config.ob[i,1])
        marker.pose.position.z = 0.0
        marker.pose.orientation.x = 0.0
        marker.pose.orientation.y = 0.0
        marker.pose.orientation.z = 0.0
        marker.pose.orientation.w = 1.0
        marker.scale.x = 0.5
        marker.scale.y = 0.5
        marker.scale.z = 1.0
        marker.color.a = 1.0
        marker.color.r = 1.0
        marker.color.g = 0.0
        marker.color.b = 0.0
        vel_marker.header.frame_id = "odom"
        vel_marker.type = Marker.ARROW
        vel_marker.action = Marker.ADD
        vel_marker.ns = "velocities"
        vel_marker.id = i
        vel_marker.pose.position.x = float(self.dwa.config.ob[i,0])
        vel_marker.pose.position.y = float(self.dwa.config.ob[i,1])
        vel_marker.pose.position.z = 0.5
        yaw = np.arctan2(self.dwa.config.ob[i,3], self.dwa.config.ob[i,2])
        # convert yaw to quaternion
        q = quaternion_from_euler(0, 0, yaw)
        vel_marker.pose.orientation.x = q[0]
        vel_marker.pose.orientation.y = q[1]
        vel_marker.pose.orientation.z = q[2]
        vel_marker.pose.orientation.w = q[3]
        vel_marker.scale.x = float(np.hypot(self.dwa.config.ob[i,2], self.dwa.config.ob[i,3])) +0.00001
        vel_marker.scale.y = 0.1
        vel_marker.scale.z = 0.1
        vel_marker.color.a = 1.0
        vel_marker.color.r = 0.0
        vel_marker.color.g = 1.0
        vel_marker.color.b = 0.0
        marker_array.markers.append(marker)
        vel_marker_array.markers.append(vel_marker)
    self.ob_pub.publish(marker_array)
    self.ob_pub.publish(vel_marker_array)

  # ros2 subscriber for odom
  def odom_callback(self, msg):

    x = msg.pose.pose.position.x
    y = msg.pose.pose.position.y
    _,_,yaw = euler_from_quaternion([msg.pose.pose.orientation.x, msg.pose.pose.orientation.y, msg.pose.pose.orientation.z, msg.pose.pose.orientation.w])
    self.current_pose = [x, y, yaw]
    self.dwa.set_current_pose(self.current_pose)
    vx = msg.twist.twist.linear.x
    az = msg.twist.twist.angular.z
    self.dwa.set_cmd_vel([vx, yaw])
    # print("current_pose received", x, y, yaw)

  def pose_callback(self, msg):
    # print("pose callback")
    x = msg.pose.pose.position.x
    y = msg.pose.pose.position.y
    _,_,yaw = euler_from_quaternion([msg.pose.pose.orientation.x, msg.pose.pose.orientation.y, msg.pose.pose.orientation.z, msg.pose.pose.orientation.w])
    self.current_pose = [x, y, yaw]
    self.dwa.set_current_pose(self.current_pose)
    # print("current_pose received", x, y, yaw)
    
  # controller timer
  def controller(self):
    # print("controller called")
    if self.goal_pose != None and self.current_pose != None:

      # print("current_pose: ", self.current_pose)
      # TODO: goal is far away break in small goal lists based on dubins path
      
      self.dwa.set_goal_pose(self.goal_pose)
      u, predicted_trajectory = self.dwa.dwa_control()
      # self.dwa.set_cmd_vel(u)
      trajectory_vis = Path()

      trajectory_vis.header.frame_id = "map"
      # trajectory_vis.header.stamp = rclpy.get_clock().now().to_msg()
      for i in range(len(predicted_trajectory)):
          pose = PoseStamped()
          pose.header.frame_id = "map"
          pose.pose.position.x = predicted_trajectory[i][0]
          pose.pose.position.y = predicted_trajectory[i][1]
          trajectory_vis.poses.append(pose)
      self.trajectory_pub.publish(trajectory_vis)
      trajectories_msg = MarkerArray() 
      trajectory_msg = Marker()

      trajectories = self.dwa.get_trajectories()

      for i in range(len(trajectories)):
          trajectory_msg = Marker()
          trajectory_msg.header.frame_id = "map"
          trajectory_msg.ns = "trajectories"
          trajectory_msg.id = i
          trajectory_msg.type = Marker.LINE_STRIP
          trajectory_msg.action = Marker.ADD
          trajectory_msg.scale.x = 0.05
          trajectory_msg.color.a = 1.0
          trajectory_msg.color.r = 0.0
          trajectory_msg.color.g = 0.0
          trajectory_msg.color.b = 1.0
          for j in range(len(trajectories[i])):
              pose = Point()
              pose.x = trajectories[i][j][0]
              pose.y = trajectories[i][j][1]
              pose.z = 0.0
              trajectory_msg.points.append(pose)
          trajectories_msg.markers.append(trajectory_msg)
      self.trajectories_pub.publish(trajectories_msg)
      
      
      vel = Twist()
      vel.linear.x = u[0]
      vel.angular.z = u[1]
      print("vel: ", vel.linear.x, vel.angular.z)
      print("goal_pose: ", self.goal_pose)
      self.publish_vel(u)

      dist_to_goal = np.hypot(self.goal_pose[0] - self.current_pose[0], self.goal_pose[1] -self.current_pose[1])
      if dist_to_goal < 0.5:
          print("goal reached")
          self.goal_pose = None
          self.publish_vel([0.0, 0.0])
    else:
      print("waiting for goal_pose", self.goal_pose, self.current_pose)
  
  def plotter(self):
    if self.goal_pose != None and self.current_pose != None:
       self.dwa.plot_all()
       
  # publish velocity
  def publish_vel(self, vel):
      msg = Twist()
      msg.linear.x = vel[0]
      msg.angular.z = vel[1]
      self.cmd_vel_pub.publish(msg)

def main(args=None):
      
      rclpy.init(args=args)
      node = rclpy.create_node('itav_agv_controller')
      controller = ITAVController(node)
      rclpy.spin(node)

if __name__ == '__main__':
    main()